wasm/validation.cpp

```diff
@@ -0,0 +1,594 @@
+}
+
+void InstValidator::push_vals(std::vector<ValueType> const &vals) {
+    value_stack.reserve(value_stack.size() + vals.size());
+
+    for (ValueType const &val : vals) {
+        value_stack.push_back(val);
+    }
+}
+
+tl::expected<std::vector<ValueOrUnknown>, ValidationError> InstValidator::pop_vals(std::vector<ValueType> const &vals) {
+    std::vector<ValueOrUnknown> popped;
+
+    // TODO(dzero): switch to std::ranges::reverse_view once we drop older toolchains
+    // NOLINTNEXTLINE(modernize-loop-convert)
+    for (auto it = vals.rbegin(); it != vals.rend(); it++) {
+        auto maybe_val = pop_val_expect(*it);
+
+        if (!maybe_val.has_value()) {
+            return tl::unexpected{maybe_val.error()};
+        }
+
+        popped.insert(popped.begin(), *maybe_val);
+    }
+
+    return popped;
+}
+
+void InstValidator::push_ctrl(Instruction i, std::vector<ValueType> params, std::vector<ValueType> results) {
+    if (!params.empty()) {
+        push_vals(params);
+    }
+
+    control_stack.push_back(
+            ControlFrame{std::move(i), std::move(params), std::move(results), value_stack.size(), false});
+    // TODO(dzero): Replace above call with below call when clang >= 16 supported in macos
+    // control_stack.emplace_back(std::move(i), std::move(params), std::move(results), value_stack.size(), false);
+}
+
+tl::expected<ControlFrame, ValidationError> InstValidator::pop_ctrl() {
+    if (control_stack.empty()) {
+        return tl::unexpected{ValidationError::ControlStackEmpty};
+    }
+
+    auto frame = control_stack.back();
+
+    if (!frame.results.empty()) {
+        auto maybe_vals = pop_vals(frame.results);
+
+        if (!maybe_vals.has_value()) {
+            return tl::unexpected{maybe_vals.error()};
+        }
+    }
+
+    if (value_stack.size() != frame.stack_height) {
+        return tl::unexpected{ValidationError::ValueStackHeightMismatch};
+    }
+
+    control_stack.pop_back();
+
+    return frame;
+}
+
+std::vector<ValueType> const &InstValidator::label_types(ControlFrame const &frame) {
+    if (std::holds_alternative<Loop>(frame.i)) {
+        return frame.params;
+    }
+
+    return frame.results;
+}
+
+void InstValidator::mark_unreachable() {
+    value_stack.resize(control_stack.back().stack_height);
+
+    control_stack.back().unreachable = true;
+}
+
+namespace {
+// TODO(dzero): Serialize operand stack and control stack as part of the ValidationError to make debugging easier
+// https://webassembly.github.io/spec/core/valid/instructions.html#instruction-sequences
+tl::expected<void, ValidationError> validate_function(std::uint32_t func_idx, Module const &m) {
+    assert(m.function_section.has_value());
+    assert(m.type_section.has_value());
+    assert(m.code_section.has_value());
+    assert(func_idx < m.function_section->type_indices.size());
+    assert(func_idx < m.code_section->entries.size());
+    assert(m.function_section->type_indices[func_idx] < m.type_section->types.size());
```

Comment:
Can the validation of these bounds be moved in here instead of being done one level up in the call stack? That way we won't need these asserts.